fix naming of RClass when template of a shared_ptr

AndLLA opened the pull request:
fix class name in R code when mixing template and shared_ptr,
as described in : https://github.com/swig/swig/issues/2372

ojwb commented:
There's no need to merge master to the branch so soon unless there have actually been relevant changes on master (which I don't see in this case) - it just ends up burning a lot of CI time for no good reason.

Not sure if this is the right way to fix this or not.  @wsfulton Maybe you could take a quick look?

wsfulton commented:
I'm not really sure, and am reluctant to merge shared_ptr changes for R without shared_ptr testing in case there is a regression. Unfortunately there is no R testing.

We have very good shared_ptr testing in other languages and it wouldn't take long for an R programmer to convert one of these to R: [java/li_boost_shared_ptr_runme.java](https://github.com/swig/swig/blob/master/Examples/test-suite/java/li_boost_shared_ptr_runme.java) or [python/li_boost_shared_ptr_runme.py](https://github.com/swig/swig/blob/master/Examples/test-suite/python/li_boost_shared_ptr_runme.py). Could you please do that for consideration for merging into master?

AndLLA commented:
Hallo,
I added li_boost_shared_ptr to the R test suite, following the java example.
I hope it works properly in CI.

In particular, regarding to this pull request:
- this pull request fixes the templates test, which fails in trunk due to the mentioned naming issue. 
Below you can find the error you would get out of trunk, showing the missing `_` after `PairT`
```
Error in getClass(Class, where = topenv(parent.frame())) : 
  “_p_SwigBoost__shared_ptrT_PairTint_doublet_t” is not a defined class
Calls: testSuite -> PairIntDouble -> new -> getClass
Execution halted
```
- there are no other differences wrt to the tests between this pull request and the trunk codebase


Note that in general when porting the test suite form java I had to disable a couple of here and there checks because:
- garbage collection is broken in R (as per bug https://www.swig.org/Doc4.0/SWIGDocumentation.html#R_nn2), therefore the final attempt at counting leaks will not work
- use_count method is not generated for everything that is not a smart_prt, (see comment `use_count not defined ...`)
- in some cases the reference counting in R generates different numbers wrt to java, I aligned it to R, maybe it's a feature ?
In the core there is nevertheless reference to the java expected reference counting.
- in a couple of cases double pointers will not work (maybe is the same issue as in #2385 ?)

In none of the above cases this pull request makes a difference,
the only thing that seems to be changing is the template test which fails for trunk and works with the proposed patch.

Thanks !!!

ojwb commented on Examples/test-suite/r/Makefile.in:
```diff
@@ -25,7 +25,8 @@ CPP_TEST_CASES += \
         r_overload_comma \
-	r_use_isnull
+	r_use_isnull \
+	li_boost_shared_ptr
```
`li_boost_shared_ptr` should already be in `CPP_TEST_CASES` (from `Examples/test-suite/common.mk`) - only R-specific testcases should need adding here.

AndLLA replied:
ok, should be fixed in the commit "fixes from code review"

ojwb commented on Source/Modules/r.cxx:
```diff
@@ -2087,9 +2087,21 @@ int R::functionWrapper(Node *n) {
-	smartname = getRClassName(smartname, 1, 1);
-	Replaceall(tm, "$R_class", smartname);
-	Delete(smartname);
+        String * classtypeobj = NewString(Getattr(parent, "classtypeobj")); // this is the correct name, e.g. ClassName(int)
```
`classtypeobj` isn't modified below so making a copy here is unnecessary.

AndLLA replied:
ok, should be fixed in the commit "fixes from code review"

wsfulton commented:
Looks like a few tests have been commented out as the implementation isn't quite right. Best to use Python as a reference for the implementation.

AndLLA commented:
Hallo,
I switched the implementation reference to python:
- most of the tests were the same, few missing added, removed a couple which had a different implementation in python
- reference numbering is now improved (i.e. R is the same as python), while a couple still different. (`# this is the python expected reference counting`)
- `use_count not defined for ...` is the same as before, here swig simply does not produce the method in several cases
- `overload_rawbyptrref` seems to be not supported at all in R

the template test, which the main point of this pull request, is the same in both implementations

Thanks

wsfulton commented:
I think the way to fix this is to add in "rtype", "scoerceout" typemaps. I have just added in these typemaps this to complete const pointers by reference support in 752b7e82cd61a80eb0f5dd80f134ba80bcae4973. I think this will fix the one runtime test that fails that does not use shared_ptr as one of the parameter or return types. Could you try testing pointerreftest again with this fix.

The way this should be done for shared_ptr is to use `$typemap`, see for example the Java equivalent to "rtype" (which is "jstype":
https://github.com/swig/swig/blob/752b7e82cd61a80eb0f5dd80f134ba80bcae4973/Lib/java/boost_shared_ptr.i#L189

The code in r.cxx that tries to replace `$R_class" can then be removed. So the implementation for getting this type correct switches from hard coding it to typemaps. That's how SWIG should work!

AndLLA commented:
Hallo,
the commit https://github.com/swig/swig/commit/752b7e82cd61a80eb0f5dd80f134ba80bcae4973 indeed fixes that test case.

Regarding the fix for templates, I tried to play around with the suggested approach but nothing seems to work.
Maybe someone more familiar with the inner workings of swig could help :smile: ?

Thanks

AndLLA commented:
Hallo,
after reviewing the code in `r.cxx` i noticed that there is another (working) implementation of the smartname to rclass algorithm.
after aligning them, the template test is a success and the bug seems to be fixed,

the two implementation are marked by a comment
```
// SmartName handling - has to be aligned to the other implementation in this file
```

wsfulton commented:
I think I can make the appropriate changes so that the R_class substitutions come from typemaps and I'll look at that. I'm confused though, the li_boost_shared_ptr test fails with me though as follows:
```
> # actually do the tests
> for (i in 1:10) {
+   print(paste("Start Loop: ", i))
+   testSuite()
+   print(paste("End Loop: ", i))
+ }
[1] "Start Loop:  1"
++++++
calling Klass
Klass(string) [me oh my]
      ++xxxxx Klass::increment tot: 2
------
++++++
calling getValue
------
[1] "PASS"
Error in f(...) : could not find function "f"
```

Are you expecting the test to pass? Could you merge master into your branch and resolve the conflict, then we can see the test failure on Github Actions. 

> after aligning them, the template test is a success and the bug seems to be fixed

There are still some parts of the test not running as they have `if (FALSE)`. Did you overlook those or are they not working?

AndLLA commented:
Hallo,
I just merged master and `li_boost_shared_ptr_runme.R` works ok either directly in R or via `make li_boost_shared_ptr.cpptest` or with `make all-r-test-suite`. 
The test `li_boost_shared_ptr` should pass with the code from this pull request (otherwise it will fails at the `templates` section if run against master).

For example, this is the output of `make all-r-test-suite`:

![image](https://user-images.githubusercontent.com/44858649/196793962-8aa0d0f7-65ba-4db0-af7b-9a34d64de14b.png)


The tests marked with `if (FALSE)` are there because they were present in the reference python implementation, so I ported them anyway.
They do not work and the reason in marked in the comments (it looks to me as unrelated to the fix proposed here):
```
undefined class _p_p_SwigBoost__shared_ptrT_Space__Klass_t
```
These tests fail anyway either against master or this pull request, therefore I left them disabled.


Thanks

AndLLA commented:
p.s.
the R tests from github actions seem to be successful:
https://github.com/swig/swig/actions/runs/3284549742/jobs/5410645458

![image](https://user-images.githubusercontent.com/44858649/196796342-38c86cfd-122a-4db1-81e9-5a2ca6b59dd8.png)


![image](https://user-images.githubusercontent.com/44858649/196797226-28522029-9d63-41f2-82be-9513d69110c8.png)

wsfulton commented:
That looks good. Ignore the `could not find function "f"` error, it was due to a local change of mine that should not have been there, sorry. I do get this though:
```
calling ~Klass
------
Error in gc(verbose = F, full = T) : unused argument (full = T)
Execution halted
~Klass() [global reference]
      --xxxxx Klass::decrement tot: 252
Makefile:49: recipe for target 'li_boost_shared_ptr.cpptest' failed
make: *** [li_boost_shared_ptr.cpptest] Error 1
william@margay:~/swig/github/swig/Examples/test-suite/r (rfix *)$ R --version
R version 3.4.4 (2018-03-15) -- "Someone to Lean On"
```
Github Actions uses this version:
```
R version 4.2.1 (2022-06-23) -- "Funny-Looking Kid"
```
It passes by dropping the `full = T` option to gc.  What is the best way to successfully call the GC for this range of versions of R? BTW, is this GC actually needed here? I commented out the loop that does GC and the test also passed with my version of R.

AndLLA commented:
Hallo,
`full = TRUE` was introduced in R 3.5.0 (released april 2018):
https://cran.r-project.org/doc/manuals/r-devel/NEWS.3.html

The idea is to force a more comprehensive collection in recent versions of R:
```
full logical; if TRUE a full collection is performed; otherwise only more recently allocated objects may be collected.
```

Since GC is anyway broken for the code produced by swig (in some cases), as described below, 
I think we can comment out the loop, forget about gc, and run the tests just once.

https://www.swig.org/Doc4.0/SWIGDocumentation.html#R_nn2

However maybe is anyway interesting to test the de-allocation as well, 
in this case we can try/catch the gc invocation in order to support older R.


Thanks

wsfulton commented:
I've made a few commits to get this into master.

1. 1d73341aa424ebcf39caab86defff5ea8f0ee5ac fixes polymorphism when wrapping C++ 'struct', so it now behaves the same as 'class'. Needed for the testcase here as most classes were C++ structs.
2. 6370fab025605ccf09e2d14e220a2cb02476427d adds support for $typemap and $R_class substitution. 
3. 6c4dcbb8fe62cd8802bb683eb9fc14b32e8459c3 uses $typemap to substitute $R_class instead attempting this in the code.
4. 6c4dcbb8fe62cd8802bb683eb9fc14b32e8459c3 also the fixed test cases. This is the vast majority of them. I didn't spend long on the remaining commented out tests as they are reference counting tests and I think they might be wrong. @AndLLA, I'm no sure if these counts were ported from Java or Python and they probably need adjusting still.
5. I've commented out the gc code as suggested.
6. I don't know enough about member variable access, consider:
```r
    kmember = MemberVariables_MemberValue_get(m); # m$getMemberValue();
```
What should it be if using Object Oriented code?

This code will go into swig-4.1.0, due for release NOW! If you could please double check it all in the next few hours, that would be much appreciated.

wsfulton commented:
BTW, I simplified the commits in this branch and added mine resulting in a5b4a4389c45f90378bffdfaea042a75f9748a66.

AndLLA commented:
Hallo,
I've enabled a couple of tests  in `li_boost_shared_ptr_runme.R` 
that are now working after merging master code,
there is only one `if (FALSE)` left.

Reference counting follows python implementation, 
now it works basically everywhere, 
with the exception of few cases that I marked with a comment:
```
# --> these are the python expected counting
```

Thanks !!!!!!!!!!

wsfulton commented:
I cherry-picked your last commit as 76f004766d245d3a0a22ccd1102e0e61e45f582d. 
I've enabled more tests and made changes for the preferred member variable syntax in 1e99e4fe37c0f2334b6de1c6bc3fcd3188bc2c92.

The reference counts are the last thing to look at then. They're probably okay as the typemaps are the same as Python and differences are likely to be down to how R cleans up objects. If not, changes can go in version 4.1.1. I'll get back to this in the next few days.
